### PR TITLE
Ron azar- JOOQ omer

### PR DESCRIPTION
**Pull request by RonAzar**

### CODE REVIEW

File: src/main/kotlin/com/hibob/academy/dao/OwnerDao.kt [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+
+class OwnerDao(private val sql: DSLContext) {
+    private val owner = OwnerTable.instance
+    private val ownerDataMapper = RecordMapper<Record,OwnerData>{ record ->
+            OwnerData(
+                record[owner.ownerName],
+                record[owner.employeeId],
+                record[owner.companyId]
+            )
+    }
```
Comment (omermelamed) [inline src/main/kotlin/com/hibob/academy/dao/OwnerDao.kt:15]: Does Owner need to have an id?

File: src/main/kotlin/com/hibob/academy/dao/OwnerDao.kt [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+
+    //Create a function that retrieve all the owners from the owner table and return their name, employee Id and company Id
+    fun getAllOwners(): List<OwnerData>{
+        return sql.select(owner.ownerName,owner.employeeId,owner.companyId)
+            .from(owner)
+            .fetch(ownerDataMapper)
+    }
```
Comment (omermelamed) [inline src/main/kotlin/com/hibob/academy/dao/OwnerDao.kt:22]: When we fetch data from the database we fetch by companyId(this is what we do in bob for security reasons).
You need to get the companyId as parameter and fetch by companyId.

File: src/main/kotlin/com/hibob/academy/dao/PetDao.kt [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+
+class PetDao(private val sql: DSLContext) {
+    private val pet = PetTable.instance
+    private val petDataMapper= RecordMapper<Record, PetData> { record ->
+        PetData(
+            record[pet.petName],
+            record[pet.dateOfArrival],
+            record[pet.companyId]
+        )
+    }
```
Comment (omermelamed) [inline src/main/kotlin/com/hibob/academy/dao/PetDao.kt:16]: Does pet need an id?

File: src/main/kotlin/com/hibob/academy/dao/PetDao.kt [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+
+    //Create a function that retrieve all the pets by a given type (represented by enum) and return their name, dateOfArrival and company Id
+    fun getAllPetsByType(type: TYPE): List<PetData> {
+        return sql.select(pet.petName, pet.dateOfArrival, pet.companyId, pet.petType)  // Include petType
+            .from(pet)
+            .where(pet.petType.eq(type.name.lowercase()))
+            .fetch(petDataMapper)
+    }
```
Comment (omermelamed) [inline src/main/kotlin/com/hibob/academy/dao/PetDao.kt:24]: When we fetch data from the database we fetch by companyId(this is what we do in bob for security reasons).
You need to get the companyId as parameter and fetch by companyId.

File: src/main/kotlin/com/hibob/academy/dao/models.kt [hunk 5]
```diff
@@ -1,3 +1,19 @@
...
+import java.time.LocalDate
+
+//return their name, dateOfArrival and company Id
+data class PetData(val petName: String, val dateOfArrival: LocalDate, val companyId: Long)
+
+//return their name, employee Id and company Id.
+data class OwnerData(val ownerName: String, val employeeId: Long, val companyId: Long)
+
+//pets by a given type (represented by enum)
+enum class TYPE{
```
Comment (omermelamed) [inline src/main/kotlin/com/hibob/academy/dao/models.kt:12]: This is bad name, it doesn't say anything regarding what it is representing.
Lets think of a better name to this enum.

File: src/main/kotlin/com/hibob/academy/dao/tables.kt [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+        val  instance = PetTable()
+    }
+}
+
+
+class OwnerTable(tableName: String = "owner"): JooqTable(tableName){
+    val ownerId = createBigIntField("owner_id")
+    val ownerName = createVarcharField("name")
+    val companyId = createBigIntField("company_id")
+    val employeeId = createBigIntField("employee_id")
```
Comment (omermelamed) [inline src/main/kotlin/com/hibob/academy/dao/tables.kt:22]: Im pretty sure that employeeId should be string field, correct me if im wrong.

File: src/main/resources/db/migration/V202410091051__pets_table.sql [hunk 7]
```diff
@@ -7,7 +7,7 @@ create table pets
     date_of_arrival date default CURRENT_DATE
 );
 
-CREATE unique index idx_pets_company_id ON pets(company_id);
+CREATE index idx_pets_company_id ON pets(company_id);
```
Comment (omermelamed) [inline src/main/resources/db/migration/V202410091051__pets_table.sql:10]: Lets move it to a different pr.

File: src/test/kotlin/com/hibob/academy/dao/OwnerDaoTest.kt [hunk 8]
```diff
@@ -0,0 +1,40 @@
...
+
+
+
+    @Test
+    fun `Insert new owner to DB`() {
+        // Insert the new owner into the database
+        dao.createNewOwner(ownerRon)
+
+        // Verify that the owner has been added to the database by filtering and checking if it contains ownerRon
+        val owners = dao.getAllOwners().filter { owner ->
```
Comment (omermelamed) [inline src/test/kotlin/com/hibob/academy/dao/OwnerDaoTest.kt:32]: Why do we need to filter the response?
Why not check the response itself?

File: src/test/kotlin/com/hibob/academy/dao/OwnerDaoTest.kt [hunk 9]
```diff
@@ -0,0 +1,40 @@
...
+            owner.ownerName == ownerRon.ownerName &&
+                    owner.employeeId == ownerRon.employeeId &&
+                    owner.companyId == ownerRon.companyId
+        }
+        // Check that ownerRon is in the list
+        assertTrue(owners.isNotEmpty(), "Owner should have been added to the database")
```
Comment (omermelamed) [inline src/test/kotlin/com/hibob/academy/dao/OwnerDaoTest.kt:38]: You can compare lists instead of check the size of the list or if it empty or not.

**Review by omermelamed — COMMENTED** [review]
Nice, left few comments.
Please add more tests that cover all scenarios

File: src/main/resources/db/migration/V202410091051__pets_table.sql (hunk 7 above)
Comment (RonAzar) [reply src/main/resources/db/migration/V202410091051__pets_table.sql:10]: Got it, Add this line to SQL Branch

File: src/main/kotlin/com/hibob/academy/dao/models.kt (hunk 5 above)
Comment (RonAzar) [reply src/main/kotlin/com/hibob/academy/dao/models.kt:12]: I Change my enum class TYPE name to a more unique and representing name- PetType

File: src/main/kotlin/com/hibob/academy/dao/tables.kt (hunk 6 above)
Comment (RonAzar) [reply src/main/kotlin/com/hibob/academy/dao/tables.kt:22]: You right, my bad

File: src/test/kotlin/com/hibob/academy/dao/OwnerDaoTest.kt (hunk 8 above)
Comment (RonAzar) [reply src/test/kotlin/com/hibob/academy/dao/OwnerDaoTest.kt:32]: Got it!

File: src/main/kotlin/com/hibob/academy/dao/OwnerDao.kt (hunk 2 above)
Comment (RonAzar) [reply src/main/kotlin/com/hibob/academy/dao/OwnerDao.kt:22]: Got it!

File: src/main/kotlin/com/hibob/academy/dao/OwnerDao.kt (hunk 1 above)
Comment (RonAzar) [reply src/main/kotlin/com/hibob/academy/dao/OwnerDao.kt:15]: Yes you'r right

Comment (RonAzar) [reply src/main/kotlin/com/hibob/academy/dao/OwnerDao.kt:15]: Actually not necessary... I use my index

File: src/main/kotlin/com/hibob/academy/dao/PetDao.kt (hunk 3 above)
Comment (RonAzar) [reply src/main/kotlin/com/hibob/academy/dao/PetDao.kt:16]: My mistake
fix it!

File: src/main/kotlin/com/hibob/academy/dao/OwnerDao.kt [hunk 10]
```diff
@@ -0,0 +1,56 @@
...
+            .from(owner)
+            .where(owner.companyId.eq(companyId))
+            .fetch(ownerDataMapper)
+    }
+
+//    Add a new API that get pet id and  return the owner’s information.
+    fun getOwnerByPetId(petId: Long, companyId: Long): OwnerData?{
+        return sql.select(owner.id, owner.ownerName,owner.employeeId,owner.companyId)
+            .from(pet)
+            .join(owner).on(pet.ownerId.eq(owner.id))
```
Comment (omermelamed) [inline src/main/kotlin/com/hibob/academy/dao/OwnerDao.kt:32]: Do we want this join operation or different join?
What are the difference between all the join operations?

File: src/main/kotlin/com/hibob/academy/dao/OwnerDao.kt [hunk 11]
```diff
@@ -0,0 +1,56 @@
...
+            .and(pet.companyId.eq(companyId))
+            .fetchOne(ownerDataMapper)  // Fetch a single record
+    }
+
+    fun getOwnerByEmployeeIdAndCompanyId(employeeId: String, companyId: Long): OwnerData?{
+        return sql.select(owner.id, owner.ownerName,owner.employeeId,owner.companyId)
+            .from(owner)
+            .where(owner.companyId.eq(companyId))
+            .and(owner.employeeId.eq(employeeId))
+            .fetchOne(ownerDataMapper)
```
Comment (omermelamed) [inline src/main/kotlin/com/hibob/academy/dao/OwnerDao.kt:43]: Nice 💪🏻

File: src/main/kotlin/com/hibob/academy/dao/PetDao.kt [hunk 12]
```diff
@@ -0,0 +1,88 @@
...
+import org.jooq.RecordMapper
+
+
+class PetDao(private val sql: DSLContext) {
+    private val pet = PetTable.instance
+    private val petDataMapper= RecordMapper<Record, PetData> { record ->
+        PetData(
+            record[pet.petName],
+            record[pet.dateOfArrival],
+            record[pet.companyId],
```
Comment (omermelamed) [inline src/main/kotlin/com/hibob/academy/dao/PetDao.kt:14]: Does pet need to have an id?
What about ownerId, do we want pet to have this or not?

File: src/main/kotlin/com/hibob/academy/dao/PetDao.kt [hunk 13]
```diff
@@ -0,0 +1,88 @@
...
+    private val petAllInfoMapper = RecordMapper<Record, PetAllInfo>{ record ->
+        PetAllInfo(
+            record[pet.id],
+            record[pet.petName],
+            record[pet.dateOfArrival],
+            record[pet.companyId],
+            PetType.valueOf(record[pet.petType].uppercase()),  // Convert string to enum (assuming your DB stores it as a string)
+            record[pet.ownerId]
+        )
+    }
```
Comment (omermelamed) [inline src/main/kotlin/com/hibob/academy/dao/PetDao.kt:27]: Why do we need 2 mappers and 2 data class?
Why not use only 1 mapper?

File: src/main/kotlin/com/hibob/academy/dao/PetDao.kt [hunk 14]
```diff
@@ -0,0 +1,88 @@
...
+
+    fun getPetByDetails(petInfo: PetAllInfo): PetAllInfo? {
+        return sql.select(pet.id, pet.petName, pet.dateOfArrival, pet.companyId, pet.petType, pet.ownerId)
+            .from(pet)  // Make sure to include 'from' in the query
+            .where(pet.petName.eq(petInfo.petName))
+            .and(pet.dateOfArrival.eq(petInfo.dateOfArrival))
+            .and(pet.companyId.eq(petInfo.companyId))
+            .and(pet.petType.eq(petInfo.petType.name))
+            .fetchOne(petAllInfoMapper)  // Use 'fetchOne' since we expect one result
+    }
```
Comment (omermelamed) [inline src/main/kotlin/com/hibob/academy/dao/PetDao.kt:67]: This method doesn't really do anything, you are passing a pet to get back the same pet?

File: src/main/kotlin/com/hibob/academy/dao/PetDao.kt [hunk 15]
```diff
@@ -0,0 +1,88 @@
...
+
+    fun insertNewPet(newPet: PetAllInfo) {
+        sql.insertInto(pet)
+            .set(pet.petName, newPet.petName)
+            .set(pet.dateOfArrival, newPet.dateOfArrival)
+            .set(pet.petType, newPet.petType.name)
+            .set(pet.companyId, newPet.companyId)
+            .set(pet.ownerId, newPet.ownerId)
+            .execute()
+    }
```
Comment (omermelamed) [inline src/main/kotlin/com/hibob/academy/dao/PetDao.kt:77]: PetAllInfo already have an id, meaning it was already created. Maybe we can use a different data class or parameters?

File: src/main/kotlin/com/hibob/academy/dao/PetDao.kt [hunk 16]
```diff
@@ -0,0 +1,88 @@
...
+
+
+    fun createNewPet(newPet: PetData) {
+        sql.insertInto(pet)
+            .set(pet.petName, newPet.petName)
+            .set(pet.dateOfArrival, newPet.dateOfArrival)
+            .set(pet.petType, PetType.DOG.name)  // Set the type dynamically
+            .set(pet.companyId, newPet.companyId)
+            .execute()
+    }
```
Comment (omermelamed) [inline src/main/kotlin/com/hibob/academy/dao/PetDao.kt:87]: What will happen if we want to insert a pet with ownerId already?

**Review by omermelamed — COMMENTED** [review]
Please add more tests and cover all scenarios

File: src/main/kotlin/com/hibob/academy/dao/OwnerDao.kt (hunk 10 above)
Comment (RonAzar) [reply src/main/kotlin/com/hibob/academy/dao/OwnerDao.kt:32]: Thanks for the input! You’re right, I’ve switched to a RIGHT JOIN to ensure we get all rows from the owner table, even if the pet isn’t linked to an owner.

File: src/main/kotlin/com/hibob/academy/dao/PetDao.kt (hunk 13 above)
Comment (RonAzar) [reply src/main/kotlin/com/hibob/academy/dao/PetDao.kt]: Got it!

File: src/main/kotlin/com/hibob/academy/dao/PetDao.kt (hunk 14 above)
Comment (RonAzar) [reply src/main/kotlin/com/hibob/academy/dao/PetDao.kt]: I use this method to retrieve the pet’s id, which is generated as a serial in the database and not available during creation.

File: src/main/kotlin/com/hibob/academy/dao/PetDao.kt (hunk 16 above)
Comment (RonAzar) [reply src/main/kotlin/com/hibob/academy/dao/PetDao.kt]: Fixed

File: src/main/kotlin/com/hibob/academy/dao/PetDao.kt (hunk 15 above)
Comment (RonAzar) [reply src/main/kotlin/com/hibob/academy/dao/PetDao.kt]: I don't understand :\

File: src/main/kotlin/com/hibob/academy/dao/OwnerDao.kt [hunk 17]
```diff
@@ -0,0 +1,56 @@
...
+    //Create a new function in the DAO file that create a new owner record (if it doesn’t exist already).
+    fun createNewOwner(newOwner: OwnerData){
+        sql.insertInto(owner)
+            .set(owner.ownerName, newOwner.ownerName)
+            .set(owner.employeeId, newOwner.employeeId)
+            .set(owner.companyId, newOwner.companyId)
+            .onConflict(owner.companyId, owner.employeeId)
+            .doNothing()
+            .execute()
+    }
```
Comment (omermelamed) [inline src/main/kotlin/com/hibob/academy/dao/OwnerDao.kt:55]: OwnerData holds id meaning it was already created.
You can create a different data class called OwnerCreationaRequest and use it or use parameters instead.

File: src/main/kotlin/com/hibob/academy/dao/PetDao.kt [hunk 18]
```diff
@@ -0,0 +1,71 @@
...
+
+    fun getPetByDetails(petInfo: PetData): PetData? {
+        return sql.select(pet.id, pet.petName, pet.dateOfArrival, pet.companyId, pet.petType, pet.ownerId)
+            .from(pet)  // Make sure to include 'from' in the query
+            .where(pet.petName.eq(petInfo.petName))
+            .and(pet.dateOfArrival.eq(petInfo.dateOfArrival))
+            .and(pet.companyId.eq(petInfo.companyId))
+            .and(pet.petType.eq(petInfo.petType.name))
+            .fetchOne(petDataMapper)  // Use 'fetchOne' since we expect one result
+    }
```
Comment (omermelamed) [inline src/main/kotlin/com/hibob/academy/dao/PetDao.kt:60]: Does this function really do anything? You are sending petData to get the same data?

File: src/main/kotlin/com/hibob/academy/dao/PetDao.kt [hunk 19]
```diff
@@ -0,0 +1,71 @@
...
+
+    fun insertNewPet(newPet: PetData) {
+        sql.insertInto(pet)
+            .set(pet.petName, newPet.petName)
+            .set(pet.dateOfArrival, newPet.dateOfArrival)
+            .set(pet.petType, newPet.petType.name)
+            .set(pet.companyId, newPet.companyId)
+            .set(pet.ownerId, newPet.ownerId)
+            .execute()
+    }
```
Comment (omermelamed) [inline src/main/kotlin/com/hibob/academy/dao/PetDao.kt:70]: Same comment as the insertOwner.
If you use PetData meaning the pet already created since it holds an id. You can use dedicated data class for creation.

**Review by omermelamed — COMMENTED** [review]
Please add more tests that cover all scenarios

File: src/main/kotlin/com/hibob/academy/dao/PetDao.kt (hunk 19 above)
Comment (RonAzar) [reply src/main/kotlin/com/hibob/academy/dao/PetDao.kt:70]: I dont think that new data class is necessary... I removed the second data class after the last review

File: src/main/kotlin/com/hibob/academy/dao/OwnerDao.kt (hunk 17 above)
Comment (RonAzar) [reply src/main/kotlin/com/hibob/academy/dao/OwnerDao.kt:55]: Got it!

File: src/main/kotlin/com/hibob/academy/dao/PetDao.kt (hunk 15 above)
Comment (RonAzar) [reply src/main/kotlin/com/hibob/academy/dao/PetDao.kt]: Fixed

File: src/main/kotlin/com/hibob/academy/dao/OwnerDao.kt [hunk 20]
```diff
@@ -0,0 +1,58 @@
...
+            .from(owner)
+            .where(owner.companyId.eq(companyId))
+            .fetch(ownerDataMapper)
+    }
+
+//    Add a new API that get pet id and  return the owner’s information.
+    fun getOwnerByPetId(petId: Long, companyId: Long): OwnerData?{
+        return sql.select(owner.id, owner.ownerName,owner.employeeId,owner.companyId)
+            .from(pet)
+            .rightJoin(owner).on(pet.ownerId.eq(owner.id))
```
Comment (omermelamed) [inline src/main/kotlin/com/hibob/academy/dao/OwnerDao.kt:32]: Do rightJoin really the right join operation here?
What are the differences between the joins?

File: src/main/kotlin/com/hibob/academy/dao/OwnerDao.kt [hunk 21]
```diff
@@ -0,0 +1,58 @@
...
+    fun insertNewOwner(ownerName: String, employeeId: String, companyId: Long): Long{
+        return sql.insertInto(owner)
+            .set(owner.ownerName, ownerName)
+            .set(owner.employeeId,employeeId)
+            .set(owner.companyId, companyId)
+            .onConflict(owner.companyId, owner.employeeId)
+            .doNothing()
+            .returning(owner.id)  // Return the generated ID after insertion
+            .fetchOne()         // Fetch the newly created row
+            ?.get(owner.id) ?: -1    // Extract the ID from the row
```
Comment (omermelamed) [inline src/main/kotlin/com/hibob/academy/dao/OwnerDao.kt:56]: Why not !! ?
If insert then for sure you will have an id.

File: src/main/kotlin/com/hibob/academy/dao/PetDao.kt [hunk 22]
```diff
@@ -0,0 +1,76 @@
...
+fun updatePetOwnerId(petId: Long, petOwnerId: Long, companyId: Long): Int {
+    // Perform the update only if ownerId is null and return the updated pet
+    val rowsAffected = sql.update(pet)
+        .set(pet.ownerId, petOwnerId)
+        .where(pet.companyId.eq(companyId))
+        .and(pet.id.eq(petId))
+        .and(pet.ownerId.isNull())  // Update only if ownerId is null
+        .execute()
+
+    return rowsAffected
```
Comment (omermelamed) [inline src/main/kotlin/com/hibob/academy/dao/PetDao.kt:44]: Do we really need to put this query into a val?
Can we just return the query?

File: src/main/kotlin/com/hibob/academy/dao/PetDao.kt [hunk 23]
```diff
@@ -0,0 +1,76 @@
...
+    fun insertNewPet(petName: String, petDateOfArrival: LocalDate, petType: PetType, companyId: Long, ownerId: Long?): Long {
+        return sql.insertInto(pet)
+            .set(pet.petName, petName)
+            .set(pet.dateOfArrival, petDateOfArrival)
+            .set(pet.petType, petType.name)
+            .set(pet.companyId, companyId)
+            .set(pet.ownerId, ownerId)
+            .returning(pet.id)  // Return the generated ID after insertion
+            .fetchOne()         // Fetch the newly created row
+            ?.get(pet.id) ?: -1    // Extract the ID from the row
```
Comment (omermelamed) [inline src/main/kotlin/com/hibob/academy/dao/PetDao.kt:74]: Same comment as insertOwner

**Review by omermelamed — COMMENTED** [review]
Please add even more tests that cover all functions and all scenarios.
Separate the resource files into a different pr.
Dao files and their tests should be in a separate pr just like migrations.

File: src/main/kotlin/com/hibob/academy/dao/OwnerDao.kt (hunk 20 above)
Comment (RonAzar) [reply src/main/kotlin/com/hibob/academy/dao/OwnerDao.kt]: Oops my bad!
I should use a left join because it ensures that the query returns the pet even if it doesn’t have an associated owner, which matches your goal of always retrieving the pet’s information.

File: src/main/kotlin/com/hibob/academy/dao/PetDao.kt (hunk 22 above)
Comment (RonAzar) [reply src/main/kotlin/com/hibob/academy/dao/PetDao.kt:44]: I did it just to clarify to myself

File: src/main/kotlin/com/hibob/academy/dao/PetDao.kt (hunk 23 above)
Comment (RonAzar) [reply src/main/kotlin/com/hibob/academy/dao/PetDao.kt:74]: I return -1 if insertion failed

**Review by omermelamed — APPROVED** [review]
